File: backend/src/movies/movie.model.js
/**
 * Movie Model
 * Firestore schema + CRUD for channel-owned long-form movies.
 *
 * Collection: channel_movies
 */

const crypto = require('crypto');
const { getFirestore } = require('../utils/firestore');

const COLLECTION = 'channel_movies';

const CONTENT_RATING_FIELDS = [
  'age_classification',
  'has_explicit_language',
  'has_nudity',
  'has_violence',
  'has_revealing_clothes',
  'has_partial_nudity',
  'has_explicit_content',
  'has_parental_guidance',
  'has_erotic_dancing',
  'has_sexual_nature',
  'has_sex',
];

function create({
  channelId,
  createdBy,
  title,
  synopsis,
  posterUrl,
  releaseDate,
  ageClassification,
  contentRating = {},
  videoSourceMode,
  hostedUrl,
  externalUrl,
  embedUrl,
  hlsUrl,
  downloadable,
}) {
  const db = getFirestore();
  const id = crypto.randomUUID();
  const now = Date.now();

  const movie = {
    id,
    channel_id: channelId,
    created_by: createdBy,
    title: title.trim(),
    synopsis: synopsis || '',
    poster_url: posterUrl || null,
    release_date: releaseDate || null,
    age_classification: ageClassification || 'teen',
    has_explicit_language: !!contentRating.has_explicit_language,
    has_nudity: !!contentRating.has_nudity,
    has_violence: !!contentRating.has_violence,
    has_revealing_clothes: !!contentRating.has_revealing_clothes,
    has_partial_nudity: !!contentRating.has_partial_nudity,
    has_explicit_content: !!contentRating.has_explicit_content,
    has_parental_guidance: !!contentRating.has_parental_guidance,
    has_erotic_dancing: !!contentRating.has_erotic_dancing,
    has_sexual_nature: !!contentRating.has_sexual_nature,
    has_sex: !!contentRating.has_sex,
    video_source_mode: videoSourceMode || 'hosted', // hosted|external_url|embed|hls
    hosted_url: hostedUrl || null,
    external_url: externalUrl || null,
    embed_url: embedUrl || null,
    hls_url: hlsUrl || null,
    downloadable: !!downloadable,
    duration: 0,
    status: 'draft', // draft|published|archived
    published_at: null,
    created_at: now,
    updated_at: now,
    total_views: 0,
    total_downloads: 0,
  };

  return db.collection(COLLECTION).doc(id).set(movie).then(() => movie);
}

async function findById(id) {
  const db = getFirestore();
  const doc = await db.collection(COLLECTION).doc(id).get();
  if (!doc.exists) return null;
  return { ...doc.data(), id: doc.id };
}

const ALLOWED_UPDATE_FIELDS = [
  'title',
  'synopsis',
  'poster_url',
  'release_date',
  'age_classification',
  ...CONTENT_RATING_FIELDS.filter((f) => f !== 'age_classification'),
  'video_source_mode',
  'hosted_url',
  'external_url',
  'embed_url',
  'hls_url',
  'downloadable',
  'duration',
  'status',
  'published_at',
];

async function update(id, fields) {
  const movie = await findById(id);
  if (!movie) return null;

  const updates = {};
  for (const key of ALLOWED_UPDATE_FIELDS) {
    if (fields[key] !== undefined) {
      updates[key] = fields[key];
    }
  }
  updates.updated_at = Date.now();

  const db = getFirestore();
  await db.collection(COLLECTION).doc(id).set(updates, { merge: true });
  return { ...movie, ...updates };
}

async function remove(id) {
  const db = getFirestore();
  await db.collection(COLLECTION).doc(id).delete();
  return true;
}

async function listByChannel(channelId, { onlyPublished = false } = {}) {
  const db = getFirestore();
  const snapshot = await db.collection(COLLECTION).where('channel_id', '==', channelId).get();
  let items = snapshot.docs.map((doc) => ({ ...doc.data(), id: doc.id }));
  if (onlyPublished) {
    items = items.filter((m) => m.status === 'published');
  }
  return items.sort((a, b) => (b.created_at || 0) - (a.created_at || 0));
}

async function listPublicFeed({ page = 1, limit = 20, publicChannelIds = [] } = {}) {
  const db = getFirestore();
  if (publicChannelIds.length === 0) {
    return { items: [], total: 0, pages: 1 };
  }

  // Firestore 'in' queries are limited to 30 values; chunk if needed.
  const chunks = [];
  for (let i = 0; i < publicChannelIds.length; i += 30) {
    chunks.push(publicChannelIds.slice(i, i + 30));
  }

  const results = await Promise.all(
    chunks.map((chunk) =>
      db
        .collection(COLLECTION)
        .where('channel_id', 'in', chunk)
        .where('status', '==', 'published')
        .get()
    )
  );

  let items = [];
  for (const snap of results) {
    items = items.concat(snap.docs.map((doc) => ({ ...doc.data(), id: doc.id })));
  }
  items.sort((a, b) => (b.published_at || b.created_at || 0) - (a.published_at || a.created_at || 0));

  const total = items.length;
  const offset = (page - 1) * limit;
  const paged = items.slice(offset, offset + limit);

  return { items: paged, total, pages: Math.max(1, Math.ceil(total / limit)) };
}

module.exports = {
  COLLECTION,
  CONTENT_RATING_FIELDS,
  create,
  findById,
  update,
  remove,
  listByChannel,
  listPublicFeed,
};

File: lib/features/vod/models/movie_model.dart
class MovieModel {
  final String id;
  final String channelId;
  final String? createdBy;
  final String title;
  final String? synopsis;
  final String? posterUrl;
  final int? releaseDate;
  final String ageClassification;

  // Content rating
  final bool hasExplicitLanguage;
  final bool hasNudity;
  final bool hasViolence;
  final bool hasRevealingClothes;
  final bool hasPartialNudity;
  final bool hasExplicitContent;
  final bool hasParentalGuidance;
  final bool hasEroticDancing;
  final bool hasSexualNature;
  final bool hasSex;

  final String videoSourceMode; // 'hosted' | 'external_url' | 'embed' | 'hls'
  final String? hostedUrl;
  final String? externalUrl;
  final String? embedUrl;
  final String? hlsUrl;
  final bool downloadable;
  final int duration; // seconds
  final String status;
  final int? publishedAt;
  final int createdAt;
  final int? updatedAt;
  final int totalViews;
  final int totalDownloads;

  const MovieModel({
    required this.id,
    required this.channelId,
    this.createdBy,
    required this.title,
    this.synopsis,
    this.posterUrl,
    this.releaseDate,
    this.ageClassification = 'teen',
    this.hasExplicitLanguage = false,
    this.hasNudity = false,
    this.hasViolence = false,
    this.hasRevealingClothes = false,
    this.hasPartialNudity = false,
    this.hasExplicitContent = false,
    this.hasParentalGuidance = false,
    this.hasEroticDancing = false,
    this.hasSexualNature = false,
    this.hasSex = false,
    this.videoSourceMode = 'hosted',
    this.hostedUrl,
    this.externalUrl,
    this.embedUrl,
    this.hlsUrl,
    this.downloadable = false,
    this.duration = 0,
    this.status = 'draft',
    this.publishedAt,
    required this.createdAt,
    this.updatedAt,
    this.totalViews = 0,
    this.totalDownloads = 0,
  });

  /// Best public playback URL for this movie, or null if none available.
  String? get playbackUrl {
    switch (videoSourceMode) {
      case 'hls':
        return hlsUrl;
      case 'external_url':
        return externalUrl;
      case 'embed':
        return embedUrl;
      case 'hosted':
      default:
        return hostedUrl;
    }
  }

  bool get isPublished => status == 'published';

  factory MovieModel.fromJson(Map<String, dynamic> json) {
    return MovieModel(
      id: _asString(json['id']),
      channelId: _asString(json['channel_id']),
      createdBy: _asNullableString(json['created_by']),
      title: _asString(json['title']),
      synopsis: _asNullableString(json['synopsis']),
      posterUrl: _asNullableString(json['poster_url']),
      releaseDate: (json['release_date'] as num?)?.toInt(),
      ageClassification: _asString(json['age_classification'], fallback: 'teen'),
      hasExplicitLanguage: _asBool(json['has_explicit_language']),
      hasNudity: _asBool(json['has_nudity']),
      hasViolence: _asBool(json['has_violence']),
      hasRevealingClothes: _asBool(json['has_revealing_clothes']),
      hasPartialNudity: _asBool(json['has_partial_nudity']),
      hasExplicitContent: _asBool(json['has_explicit_content']),
      hasParentalGuidance: _asBool(json['has_parental_guidance']),
      hasEroticDancing: _asBool(json['has_erotic_dancing']),
      hasSexualNature: _asBool(json['has_sexual_nature']),
      hasSex: _asBool(json['has_sex']),
      videoSourceMode: _asString(json['video_source_mode'], fallback: 'hosted'),
      hostedUrl: _asNullableString(json['hosted_url']),
      externalUrl: _asNullableString(json['external_url']),
      embedUrl: _asNullableString(json['embed_url']),
      hlsUrl: _asNullableString(json['hls_url']),
      downloadable: _asBool(json['downloadable']),
      duration: (json['duration'] as num?)?.toInt() ?? 0,
      status: _asString(json['status'], fallback: 'draft'),
      publishedAt: (json['published_at'] as num?)?.toInt(),
      createdAt: (json['created_at'] as num?)?.toInt() ?? 0,
      updatedAt: (json['updated_at'] as num?)?.toInt(),
      totalViews: (json['total_views'] as num?)?.toInt() ?? 0,
      totalDownloads: (json['total_downloads'] as num?)?.toInt() ?? 0,
    );
  }

  static String _asString(dynamic value, {String fallback = ''}) {
    if (value == null) return fallback;
    return value.toString();
  }

  static String? _asNullableString(dynamic value) {
    if (value == null) return null;
    final parsed = value.toString();
    return parsed.isEmpty ? null : parsed;
  }

  static bool _asBool(dynamic value) {
    if (value is bool) return value;
    if (value is num) return value != 0;
    if (value is String) {
      final normalized = value.trim().toLowerCase();
      if (normalized == 'true' || normalized == '1') return true;
      if (normalized == 'false' || normalized == '0') return false;
    }
    return false;
  }
}
